### PR TITLE
Ex 02

### PR DESCRIPTION
**Pull request by vsnupoudel**
Added the codes in. Tried to make codes more compact as per suggestions from the previous review.

### CODE REVIEW

**Comment by vsnupoudel** [comment]
Hi Ngoc,
I accidently merged yngve's  remote master with my ex_02. Can I fix it after your review.

**Comment by vsnupoudel** [comment]
> Hi Ngoc,
> I accidently merged yngve's remote master with my ex_02. Can I fix it after your review.
> I will revert the branch back to a previous commit.

It seems I was able to force it to a previous commit now. Please let me know if I need to correct something else.

File: src/bishnu_poudel_ex/ex02/bubble_sort.py [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
-        return data
+    data2 = list(data1)
+    end = len(data2)
+    if len(data2) <= 1:
+        return data2
     else:
-        for j in range(len(data)):
+        for j in range(len(data2)):
             end = end - 1
             for i in range(end):
```
Comment (huynhngoc) [inline src/bishnu_poudel_ex/ex02/bubble_sort.py]: Maybe `for i in range(len(data2) - j - 1`, then you won't need a tracking `end`

File: src/bishnu_poudel_ex/ex02/bubble_sort.py [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
-                    data[i + 1] = data[i]
-                    data[i] = temp
-    return data
-
-
-# print( bubble_sort([1,2,4,2,1]) )
+                if data2[i + 1] < data2[i]:
+                    temp = data2[i + 1]
+                    data2[i + 1] = data2[i]
+                    data2[i] = temp
```
Comment (huynhngoc) [inline src/bishnu_poudel_ex/ex02/bubble_sort.py]: Ever heard of python swapping `i, j = j, i`? Super short and easy.

File: src/bishnu_poudel_ex/ex02/message_entropy.py [hunk 3]
```diff
@@ -1,28 +1,34 @@
...
 
 def entropy(message):
-    result_freq=letter_freq(message)  # The function above returns a dictionary
-    sum_values=0
-    for i in result_freq.values():  # taking the sum of all the values. i.e. finding N- total characters in the text
-        sum_values=sum_values+i
+    result_freq = letter_freq(message)
+    sum_values = 0
+    for i in result_freq.values():
+        sum_values = sum_values + i
```
Comment (huynhngoc) [inline src/bishnu_poudel_ex/ex02/message_entropy.py]: `sum_values = len(message)` is shorter and easiert.

File: src/bishnu_poudel_ex/ex02/message_entropy.py [hunk 4]
```diff
@@ -1,28 +1,34 @@
...
     entr = 0
     for i in result_freq.values():
-        entr=entr+ (i/sum_values) *math.log((i/sum_values),2)  # multiplying probability with its log 2 value and summing up
+        entr = entr + (i / sum_values) * math.log((i / sum_values),
```
Comment (huynhngoc) [inline src/bishnu_poudel_ex/ex02/message_entropy.py]: Ever heard of `math.log2`?

File: src/bishnu_poudel_ex/ex02/bubble_sort.py (hunk 1 above)
Comment (vsnupoudel) [reply src/bishnu_poudel_ex/ex02/bubble_sort.py]: Done. Thank you!

File: src/bishnu_poudel_ex/ex02/bubble_sort.py (hunk 2 above)
Comment (vsnupoudel) [reply src/bishnu_poudel_ex/ex02/bubble_sort.py]: Done. That was easy! Thanks

File: src/bishnu_poudel_ex/ex02/message_entropy.py (hunk 3 above)
Comment (vsnupoudel) [reply src/bishnu_poudel_ex/ex02/message_entropy.py]: Got you. I did it this way, as I previously thought we do not need newline characters while calculating entropy. But it seems, they are included in the dictionary anyway. Changed this also.. Thanks

File: src/bishnu_poudel_ex/ex02/message_entropy.py (hunk 4 above)
Comment (vsnupoudel) [reply src/bishnu_poudel_ex/ex02/message_entropy.py]: Done. Thank you!

**Pull request merged**